### DISCUSSION TITLE
운영체제 - CPU 스케줄링 - 호현&일구&성은

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#123] [질의 응답]
- CPU 스케쥴러는 무엇인가요?
- CPU 스케줄링의 성능 척도에는 어떤 것들이 있나요?
- 운영체제가 스케줄링 결정을 내리는 시점과 방식에 대해 설명해주세요.
- 스케줄링의 목적과 저, 중, 고수준 단계에 대해 말해주세요.
- 스케줄러가 어떤 프로세스에 우선적으로 CPU를 할당할지 스케줄링 시 고려하는 사항에 대해 말해주세요.

**Comment by seongeun42** [comment]
### ❓️ CPU 스케줄러는 무엇인가요?

<br>

### 💬 답변
<blockquote>

* 다중 프로그램 OS의 기본으로, 여러 프로세스들이 CPU를 교환하며 보다 생산적으로 동작
* CPU를 선점한 프로세스가 대기하는 시간을 보다 효율적으로 사용하기 위하여 사용
* 기본적으로 다음에 실행될 프로세스를 정하고, 프로세스들을 실행가능한 상태로 만듦

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/44dc4509-e993-4bc1-988d-8fff2fba6f67)

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ CPU 스케줄링의 성능 척도에는 어떤 것들이 있나요?

<br>

### 💬 답변
<blockquote>

* CPU의 입장
   * CPU utilization(CPU 사용률)
      * 시스템의 동작 시간 중 CPU가 사용된 시간을 측정하는 방법으로 최대한 CPU를 바쁘게 만드는 것
      * 가장 이상적인 수치는 100%
   * Throughput(처리량)
      * 단위 시간당 작업을 마친 프로세스의 수다. 즉, CPU 버스트를 처리한 수
* 프로세스의  입장
   * Trun-around time(반환 시간)
      * 프로세스가 생성된 후 종료되어 사용하던 자원을 모두 반환하는 데까지 걸리는 시간
      *  프로세스의 대기 시간 + 실행시간
         *  대기 시간은 없을 수도 있고, 1 번일 수도, 여러 번일수도 있음
   * Waiting time(대기 시간)
      * 프로세스가 CPU를 할당받아 실행되기 전 대기 상태일 때의 시간
         *  보통 준비 큐에서 대기를 하는 시간
   * Response time(응답 시간)
      * 프로세스가 대기 상태에 들어와 CPU를 최초로 얻기까지 걸리는 시간
         *  대기 시간과의 차이점은 대기 시간은 반환 시간과 마찬가지로 여러 번 있을 수 있음
      *  그 총합이 대기 시간이고, 응답 시간은 최초의 한 번
         *  응답 시간은 프로세스 입장에서 CPU를 한 번도 못 얻은 것과 한 번이라도 얻는 것은 사용자 응답에 있어서 중요한 차이가 있기 때문에 중요

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 운영체제가 스케줄링 결정을 내리는 시점과 방식에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

운영체제의 스케줄링 결정 시점은 CPU를 점유한 프로세스가 다른 프로세스로 전환 때이다. 즉, 프로세스의 종료, I/O요청, 인터럽트 발, 주어진 타임 슬라이스 사용, 우선순위 변경 등의 상황이 스케줄링 시점이라고 할 수 있다.

스케줄링 방식에는 크게 선점형 스케줄링과 비선점형 스케줄링으로 나눌 수 있다. 선점형 스케줄링은 어떤 프로세스가 CPU를 할당받아 실행 중이더라도 CPU를 강제로 뺏길 수 있는 방식이고, 비선점형 스케줄링은 어떤 프로세스가 CPU를 할당받아 실행 중이라면 끝날 때까지 아무도 뺏을 수 없는 방식이다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 스케줄링의 목적과 저, 중, 고수준 단계에 대해 말해주세요.

<br>

### 💬 답변
<blockquote>

아래에 댓글로 나눠서 작성

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 스케줄링의 목적

<br>

### 💬 답변
<blockquote>

* 공평성 : 모든 프로세스가 자원을 공평하게 배정받아야 하며, 특정 프로세스가 배제되어서는 안 됨
* 효율성 : 시스템 자원을 놀리는 시간 없이 스케줄링해야 함
* 안정성 : 우선순위를 사용하여 중요한 프로세스가 먼저 처리되도록 해야 함
* 반응 시간 보장 : 응답이 없는 경우 사용자는 시스템이 멈춘 것으로 가정하기 때문에 시스템은 적절한 시간 안에 프로세스의 요구에 반응해야 함
* 무한 연기 방지 : 특정 프로세스의 작업이 무한히 연기되어서는 안 됨

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 저, 중, 고수준 단계

<br>

### 💬 답변
<blockquote>

<img src="https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/bdae8869-f2c2-4137-9027-471ec2a2536e" width="500">

* 고수준 스케줄링(long-term scheduling, job scheduling, admissin scheduling)
   * 가장 큰 틀에서 이루어지는 CPU 스케줄링으로 시스템 내의 전체 작업 수를 조절
   *  시스템 과부하를 막기 위해 어떤 작업을 시스템이 받아들일지 또는 거부할지를 결정하므로 시스템 내에서 동작 시에 실행 가능한 프로세스의 총개수가 정해짐
   * 주로 메인프레임 같은 큰 시스템에서 규모가 큰 일괄 작업을 처리할 때 사용함
* 중간 수준 스케줄링
   * 중지(suspend)와 활성화(active)로 전체 시스템의 활성화된 프로세스 수를 조절함
     * 프로세스가 활성화된 후에도 여러 요인으로 시스템 과부하가 걸릴 수 있기 때문
   * 이로 인해 저수준 스케줄링이 원만하게 이루어지도록 완충하는 역할을 함
   * 보류된 프로세스는 여유가 생기면 다시 활성화됨
* 저수준 스케줄링(short-term scheduling)
   * 가장 작은 단위의 스케줄링으로 어떤 프로세스에 CPU를 할당할지, 어떤 프로세스를 대기 상태로 보낼지 등을 결정
   * 중간 수준의 스케줄링은 프로세스를 보류 상태로 보내고, 저수준 스케줄링은 대기 상태로 보냄
   * 스케줄링에 대해 공부하는 대부분의 내용이 이 저수준 스케줄링
     * 즉, 준비 상태 <-> 실행 상태 <-> 대기 상태 등 프로세스의 상태 변화가 여기서 일어남

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 스케줄러가 어떤 프로세스에 우선적으로 CPU를 할당할지 스케줄링 시 고려하는 사항에 대해 말해주세요.

<br>

### 💬 답변
<blockquote>

* preemptive vs non-preemptive
   * 선점형 스케줄링 : 프로세스가 CPU를 할당받아 실행 중이더라도 운영체제가 CPU를 강제로 빼앗을 수 있는 방식
      * CPU 처리 시간이 매우 긴 프로세스가 CPU 사용 독점을 막을 수 있어 효율적인 운영이 가능
      * 하지만 잦은 Context Switching 으로 오버헤드가 많이 발생
   * 비선점형 스케줄링 : 프로세스가 CPU를 점유하고 있다면 이를 빼앗을 수 없는 방식
      * 필요한 문맥 교환만 일어나기 때문에 오버헤드가 상대적으로 적지만 프로세스의 배치에 따라 효율성 차이가 많이 남
* CPU bound vs I/O bound
   * 프로세스가 대기 상태에 있다가 CPU를 할당받아 실행하면 CPU burst, 입출력 작업을 하면 I/O burst
   * CPU bound process(CPU 집중 프로세스) : CPU를 많이 사용하여 CPU burst가 많은 프로세스
   * I/O bound process(입출력 집중 프로세스) : 입출력을 많이 사용해 I/O burst가 많은 프로세스
   * 두 프로세스가 같이 대기상태에 있다면 입출력 집중 프로세스에 먼저 CPU를 할당시키는 것이 더 효율적
      * 왜냐하면 입출력 집중 프로세스는 CPU를 빠르게 쓰고 입출력 버스트를 하러 나가기 때문에 다른 프로세스가 오래 기다리지 않아도 됨
* 전면 프로세스 vs 후면 프로세스
   * 전면 프로세스 : GUI를 사용하는 운영체제에서 화면의 맨 앞에 놓여 현재 입출력이 사용됨.
      *  사용자와 상호작용이 가능해 상호작용 프로세스라고도 불림 (워드 프로세스)
   * 후면 프로세스 : 사용자의 입력 없이 작동하여 일괄 작업 프로세스라고 불림 (압축 프로세스)
   * 전면 프로세스는 사용자의 요구에 즉각 즉각 반응해야 하지만 후면 프로세스는 그럴 필요가 없음
      *  따라서 전면 프로세스를 먼저 처리해 줘야 한다.
* 프로세스 우선순위
   * CPU 스케줄러 대부분은 프로세스에 우선순위를 매겨 우선순위가 높은(우선순위 숫자가 작은) 프로세스부터 처리되도록 함
   * 커널 프로세스 > 일반 프로세스
   * 전면 프로세스 > 후면 프로세스
   * 대화형 프로세스 > 일괄 처리 프로세스
   * 입출력 집중 프로세스 > CPU 집중 프로세스

</blockquote>

<br>
